docs/proposals-accepted/202304-query-path-tenancy.md

```diff
@@ -0,0 +1,139 @@
+-->
+
+Any team running Thanos in a multi-tenant environment.
+
+## Non-Goals
+
+<!--
+* Move old designs to the new format.
+* Not doing X,Y,Z.
+-->
+
+* Add multi-tenancy to Thanos Ruler. It's not needed. One will still be able to point their Ruler to the multi-tenant Query Frontend, effectively making it multi-tenant. 
+* Implement cross-tenant querying. It poses the question about how to track metrics of a multi-tenant query, probably requiring new logic for query splitting and/or separation of the concepts of query initiator tenant and query target tenants. To exemplify: if tenant A executes a query touching their own data and tenant B data (imagine a label matcher `tenant_id=~"A|B"`), how does Thanos export a request duration metric about this query? This is out of scope and an specific proposal can be created for it.
+
+## How
+
+<!--
+Explain the full overview of the proposed solution. Some guidelines:
+
+* Make it concise and **simple**; put diagrams; be concrete, avoid using “really”, “amazing” and “great” (:
+* How you will test and verify?
+* How you will migrate users, without downtime. How we solve incompatibilities?
+* What open questions are left? (“Known unknowns”)
+-->
+
+* Implement a mechanism to allow incoming requests to specify the tenant being queried using an HTTP header. This applies to all components in the query path. We follow the example of Thanos Receive, which uses the `--receive.tenant-header="THANOS-TENANT"` flag to configure the tenant header, adapting its name to each component.
+* The tenant header value from a given request should travel downstream to all the components being called, so that it can be added to their metrics, traces, and logs without requiring duplicated/extra work to re-parse the query. This also applies to gRPC calls.
+* Implement a command line argument in the Querier component that will indicate with label name that should be used to enforce tenancy. Again, we follow the example of Thanos Receive, which uses the `--receive.tenant-label-name="tenant_id"` flag to achieve this, and adapt its name. The default behavior, when the flag as an empty string as value, is to not identify tenants. 
+  * The label verification and enforcement should be done by reusing prom-label-proxy's [Enforce.EnforceMatchers](https://github.com/prometheus-community/prom-label-proxy/blob/main/injectproxy/enforce.go#L141). There's no reason to (re)implement something specific and special for Thanos.
+  * This behavior should be implemented as part of the base logic of both the HTTP and gRPC query APIs, before the query is handed to the query engine. This allows users managing complex Querier trees can choose where they want this logic to be enabled. 
+* Update metrics exported by the components in the query path to include the tenant label when it's available. 
+* Implement a tenant selector in the Query Frontend UI, which should communicate the tenant to Query Frontend using the HTTP header.
```

Comment:
I'm not entirely sure I understand this. Why have a selector in the UI? 
Likely tenants will be authenticated in some layer above Thanos, and that is where tenant HTTP headers would be added (even for cross-tenant cases which are out of scope here).

The UI would just be proxied, and would make requests with the header present.

Reply:
> Likely tenants will be authenticated in some layer above Thanos

This is completely optional, we cannot assume everyone will have authentication. One might still want to run queries with multi-tenancy checks enabled even without an authentication layer, most importantly because it enabled observability of queries based on tenant all the way throughout the query's path.

Reply:
Thinking out loud and taking the scenario where authn/authz happens before the user reaches the UI, does adding a selector here pose any risk to leaking anything?

Reply:
@PhilipGough it does. But this demands some higher level thinking: is it Thanos' responsibility that there is no data leak? IMHO, it shouldn't be.

If you get to the point of seeing a Thanos Query (Frontend or not) UI, you can easily watch you network tab and craft a request that will return data from other tenants. We can only avoid this by making Thanos authn aware, which is not something Thanos wants to do.

Reply:
I would change the wording though. In the query UI it wouldn't be a selector, but instead a free-type textbox.

Reply:
I agree. I just wanted to call out the issue